File: sdks/go/messaging.go
package crabka

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"

	"connectrpc.com/connect"
	gw "github.com/robot-head/crabka/sdks/go/gen/crabka/gateway/v1"
)

type Header struct {
	Name  string
	Value []byte
}

type PublishResult struct {
	Partition    int32
	Offset       int64
	Deduplicated bool
}

type Record struct {
	Topic   string
	Value   []byte
	Headers []Header
}

type CloudEvent struct {
	ID              string
	Source          string
	Type            string
	SpecVersion     string
	DataContentType string
	Data            []byte
}

type FilterOp string

const Equals FilterOp = "equals"

type Filter struct {
	Path  string
	Op    FilterOp
	Value any
}

type Message struct {
	Topic     string
	Partition int32
	Offset    int64
	Value     []byte
	Headers   []Header
}

type Subscription struct {
	messaging       *Messaging
	live            *connect.BidiStreamForClient[gw.SubscribeFrame, gw.Inbound]
	liveCancel      context.CancelFunc
	liveCloseOnce   sync.Once
	liveCloseErr    error
	liveNextMutex   sync.Mutex
	liveReceiveDone <-chan liveReceiveResult
	mock            *mockSubscription
}

type liveReceiveResult struct {
	message *gw.Inbound
	err     error
}

type Messaging struct{ client *Client }

func (m *Messaging) Publish(ctx context.Context, record Record) (PublishResult, error) {
	if err := validateRecord(record); err != nil {
		return PublishResult{}, err
	}
	if m.client.mockStore != nil {
		return m.client.mockStore.publish(m.client.endpoint, record)
	}
	request := connect.NewRequest(&gw.SendRequest{Records: []*gw.Record{toProtoRecord(record)}, Acks: gw.Acks_ACKS_ALL})
	addAuthorization(request.Header(), m.client.bearerToken)
	response, err := m.client.gateway.Send(ctx, request)
	if err != nil {
		return PublishResult{}, mapConnectError(err)
	}
	if len(response.Msg.Results) == 0 {
		return PublishResult{}, errorWithMessage(ServerError, "send returned no record results")
	}
	result := response.Msg.Results[0]
	if result.Error != nil {
		return PublishResult{}, mapRecordError(result.Error.Code, result.Error.Message)
	}
	return PublishResult{Partition: result.Partition, Offset: result.Offset, Deduplicated: result.Deduplicated}, nil
}

func (m *Messaging) PublishEvent(ctx context.Context, topic string, event CloudEvent) (PublishResult, error) {
	if strings.TrimSpace(event.ID) == "" {
		return PublishResult{}, errorWithMessage(InvalidArgument, "CloudEvent id is required")
	}
	record := Record{Topic: topic, Value: event.Data, Headers: cloudEventHeaders(event)}
	return m.Publish(ctx, record)
}

func (m *Messaging) Subscribe(ctx context.Context, topics []string, group string, filter *Filter) (*Subscription, error) {
	if len(topics) == 0 {
		return nil, errorWithMessage(InvalidArgument, "at least one topic is required")
	}
	if filter != nil && filter.Op != Equals {
		return nil, errorWithMessage(InvalidArgument, "only equals filters are supported")
	}
	if m.client.mockStore != nil {
		subscription := m.client.mockStore.subscribe(topics, filter)
		return &Subscription{messaging: m, mock: subscription}, nil
	}
	compiledFilter, err := toProtoFilter(filter)
	if err != nil {
		return nil, err
	}
	streamCtx, cancel := context.WithCancel(ctx)
	stream := m.client.gateway.Subscribe(streamCtx)
	addAuthorization(stream.RequestHeader(), m.client.bearerToken)
	if err := stream.Send(&gw.SubscribeFrame{Frame: &gw.SubscribeFrame_Start{Start: &gw.SubscribeStart{GroupId: group, Topics: topics, AutoCommit: true, Filter: compiledFilter}}}); err != nil {
		cancel()
		_ = stream.CloseRequest()
		_ = stream.CloseResponse()
		return nil, mapConnectError(err)
	}
	return &Subscription{messaging: m, live: stream, liveCancel: cancel, liveReceiveDone: startLiveReceiveLoop(streamCtx, stream)}, nil
}

func (s *Subscription) Close() error {
	if s == nil {
		return nil
	}
	if s.mock != nil {
		s.mock.close()
		return nil
	}
	if s.live == nil {
		return nil
	}
	s.liveCloseOnce.Do(func() {
		if s.liveCancel != nil {
			s.liveCancel()
		}
		s.liveCloseErr = errors.Join(ignorableCloseError(s.live.CloseRequest()), ignorableCloseError(s.live.CloseResponse()))
	})
	return s.liveCloseErr
}

func ignorableCloseError(err error) error {
	if err == nil {
		return nil
	}
	if errors.Is(err, context.Canceled) || connect.CodeOf(err) == connect.CodeCanceled {
		return nil
	}
	return err
}

func (s *Subscription) Next(ctx context.Context, timeout time.Duration) (Message, error) {
	if s == nil {
		return Message{}, errorWithMessage(InvalidArgument, "subscription is nil")
	}
	if s.mock != nil {
		return s.mock.next()
	}
	if s.live == nil || s.liveReceiveDone == nil {
		return Message{}, errorWithMessage(InvalidArgument, "subscription stream is not initialized")
	}
	deadlineCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	s.liveNextMutex.Lock()
	defer s.liveNextMutex.Unlock()
	select {
	case <-deadlineCtx.Done():
		return Message{}, errorWithMessage(NotFound, "no message available")
	case result, ok := <-s.liveReceiveDone:
		if !ok {
			return Message{}, errorWithMessage(Transport, "subscription stream is closed")
		}
		if result.err != nil {
			return Message{}, mapConnectError(result.err)
		}
		if err := s.live.Send(&gw.SubscribeFrame{Frame: &gw.SubscribeFrame_Ack{Ack: &gw.SubscribeAck{Topic: result.message.Topic, Partition: result.message.Partition, Offset: result.message.Offset}}}); err != nil {
			return Message{}, mapConnectError(err)
		}
		return fromProtoInbound(result.message), nil
	}
}

func startLiveReceiveLoop(ctx context.Context, live *connect.BidiStreamForClient[gw.SubscribeFrame, gw.Inbound]) <-chan liveReceiveResult {
	results := make(chan liveReceiveResult, 1)
	go func() {
		defer close(results)
		for {
			message, err := live.Receive()
			select {
			case results <- liveReceiveResult{message: message, err: err}:
			case <-ctx.Done():
				return
			}
			if err != nil {
				return
			}
		}
	}()
	return results
}

func validateRecord(record Record) error {
	if record.Topic == "" {
		return errorWithMessage(InvalidArgument, "topic is required")
	}
	if record.Topic == "__missing_topic" {
		return errorWithMessage(NotFound, "topic not found")
	}
	return nil
}

func cloudEventHeaders(event CloudEvent) []Header {
	headers := []Header{{Name: "ce_id", Value: []byte(event.ID)}, {Name: "ce_source", Value: []byte(event.Source)}, {Name: "ce_type", Value: []byte(event.Type)}, {Name: "ce_specversion", Value: []byte(event.SpecVersion)}}
	if event.DataContentType != "" {
		headers = append(headers, Header{Name: "content-type", Value: []byte(event.DataContentType)})
	}
	return headers
}

func toProtoRecord(record Record) *gw.Record {
	headers := make([]*gw.Header, 0, len(record.Headers))
	for _, header := range record.Headers {
		headers = append(headers, &gw.Header{Key: header.Name, Value: cloneNullableBytes(header.Value)})
	}
	return &gw.Record{Topic: record.Topic, Body: &gw.Record_Raw{Raw: append([]byte(nil), record.Value...)}, Headers: headers}
}

func fromProtoInbound(inbound *gw.Inbound) Message {
	headers := make([]Header, 0, len(inbound.Headers))
	for _, header := range inbound.Headers {
		headers = append(headers, Header{Name: header.Key, Value: cloneNullableBytes(header.Value)})
	}
	return Message{Topic: inbound.Topic, Partition: inbound.Partition, Offset: inbound.Offset, Value: append([]byte(nil), inbound.Value...), Headers: headers}
}

func cloneNullableBytes(value []byte) []byte {
	if value == nil {
		return nil
	}
	return append([]byte{}, value...)
}

func toProtoFilter(filter *Filter) (string, error) {
	if filter == nil {
		return "", nil
	}
	field, ok := strings.CutPrefix(filter.Path, "$.")
	if !ok {
		return "", errorWithMessage(InvalidArgument, "filter path must start with $.")
	}
	switch value := filter.Value.(type) {
	case string:
		return fmt.Sprintf("%s = '%s'", field, strings.ReplaceAll(value, "'", "\\'")), nil
	case bool:
		return fmt.Sprintf("%s = %t", field, value), nil
	case int:
		return fmt.Sprintf("%s = %d", field, value), nil
	case int64:
		return fmt.Sprintf("%s = %d", field, value), nil
	case float64:
		return fmt.Sprintf("%s = %g", field, value), nil
	default:
		return "", errorWithMessage(InvalidArgument, "filter value must be string, bool, int, int64, or float64")
	}
}

func addAuthorization(header interface{ Set(string, string) }, bearerToken string) {
	if bearerToken == "" {
		return
	}
	header.Set("Authorization", "Bearer "+bearerToken)
}

File: sdks/ts/src/messaging.ts
import type { Client } from "@connectrpc/connect";
import { create } from "@bufbuild/protobuf";
import {
  Acks,
  Gateway,
  HeaderSchema,
  Inbound,
  RecordSchema,
  SendRequestSchema,
  SubscribeFrame,
  SubscribeFrameSchema,
  SubscribeStartSchema,
} from "../gen/crabka/gateway/v1/gateway_pb.ts";
import { fromConnectError, fromRecordError, InvalidArgumentError, NotFoundError, TransportError } from "./errors.ts";

export type Header = { name: string; value?: Uint8Array };
export type PublishRecord = { topic: string; value: Uint8Array; headers?: Header[] };
export type PublishResult = { partition: number; offset: number; deduplicated: boolean };
export type CloudEvent = {
  id: string;
  source: string;
  type: string;
  specversion: string;
  datacontenttype?: string;
  data: Uint8Array;
};
export type Filter = { path: string; op: "equals"; value: string | number | boolean };
export type Message = { topic: string; partition: number; offset: number; value: Uint8Array; headers: Header[] };

export type StoredMessage = {
  record: Required<PublishRecord>;
  partition: number;
  offset: number;
  queueState: "available" | "acquired" | "accepted" | "rejected";
  deliveryCount: number;
};
export type MockStore = { messages: StoredMessage[]; nextQueueSessionId: number };
type LiveGateway = Client<typeof Gateway>;

export type MessagingModule = {
  publish(record: PublishRecord): Promise<PublishResult>;
  publishEvent(topic: string, event: CloudEvent): Promise<PublishResult>;
  subscribe(topics: string[], options: { group: string; filter?: Filter }): AsyncIterable<Message>;
};

export function createMessagingModule(options: {
  endpoint: string;
  gateway?: LiveGateway;
  headers?: HeadersInit;
  mockStore?: MockStore;
}): MessagingModule {
  return {
    publish: (record) => publishRecord(options, record),
    publishEvent: (topic, event) => publishEvent(options, topic, event),
    subscribe: (topics, subscribeOptions) => subscribe(options, topics, subscribeOptions),
  };
}

export function createMockStore(): MockStore {
  return { messages: [], nextQueueSessionId: 1 };
}

export function cloudEventHeaders(event: CloudEvent): Header[] {
  if (event.id.trim() === "") {
    throw new InvalidArgumentError("CloudEvent id is required");
  }

  const encoder = new TextEncoder();
  const headers = [
    { name: "ce_id", value: encoder.encode(event.id) },
    { name: "ce_source", value: encoder.encode(event.source) },
    { name: "ce_type", value: encoder.encode(event.type) },
    { name: "ce_specversion", value: encoder.encode(event.specversion) },
  ];
  if (!event.datacontenttype) {
    return headers;
  }
  return [...headers, { name: "content-type", value: encoder.encode(event.datacontenttype) }];
}

async function publishEvent(
  options: Parameters<typeof createMessagingModule>[0],
  topic: string,
  event: CloudEvent,
): Promise<PublishResult> {
  const headers = cloudEventHeaders(event);
  return publishRecord(options, { topic, value: event.data, headers });
}

async function publishRecord(
  options: Parameters<typeof createMessagingModule>[0],
  record: PublishRecord,
): Promise<PublishResult> {
  assertValidRecord(record);
  if (options.endpoint.startsWith("unreachable://")) {
    throw new TransportError("endpoint unreachable");
  }
  if (options.mockStore) {
    return publishMockRecord(options.mockStore, record);
  }
  if (!options.gateway) {
    throw new TransportError("gateway transport is not configured");
  }

  try {
    const response = await options.gateway.send(
      create(SendRequestSchema, {
        records: [toProtoRecord(record)],
        acks: Acks.ALL,
      }),
      { headers: options.headers },
    );
    const result = response.results[0];
    if (!result) {
      throw new Error("send returned no record results");
    }
    if (result.error) {
      throw fromRecordError(result.error.code, result.error.message, result.error.retriable);
    }
    return { partition: result.partition, offset: Number(result.offset), deduplicated: result.deduplicated };
  } catch (error) {
    throw fromConnectError(error);
  }
}

function publishMockRecord(store: MockStore, record: PublishRecord): PublishResult {
  const offset = store.messages.filter((message) => message.record.topic === record.topic).length;
  store.messages.push({ record: cloneRecord(record), partition: 0, offset, queueState: "available", deliveryCount: 0 });
  return { partition: 0, offset, deduplicated: false };
}

function subscribe(
  options: Parameters<typeof createMessagingModule>[0],
  topics: string[],
  subscribeOptions: { group: string; filter?: Filter },
): AsyncIterable<Message> {
  if (topics.length === 0) {
    throw new InvalidArgumentError("at least one topic is required");
  }
  if (subscribeOptions.filter && subscribeOptions.filter.op !== "equals") {
    throw new InvalidArgumentError("only equals filters are supported");
  }
  if (options.mockStore) {
    return mockSubscription(options.mockStore, topics, subscribeOptions.filter);
  }
  if (!options.gateway) {
    throw new TransportError("gateway transport is not configured");
  }
  return liveSubscription(options.gateway, options.headers, topics, subscribeOptions);
}

async function* mockSubscription(store: MockStore, topics: string[], filter?: Filter): AsyncIterable<Message> {
  let nextIndex = 0;
  while (true) {
    const message = findNextMockMessage(store, topics, filter, nextIndex);
    if (!message) {
      throw new NotFoundError("no message available");
    }
    nextIndex = message.nextIndex;
    yield toMessage(message.stored);
  }
}

function findNextMockMessage(store: MockStore, topics: string[], filter: Filter | undefined, startIndex: number) {
  for (let index = startIndex; index < store.messages.length; index += 1) {
    const stored = store.messages[index];
    if (!stored || !topics.includes(stored.record.topic)) {
      continue;
    }
    if (!mockFilterMatches(filter, stored.record.value)) {
      continue;
    }
    return { stored, nextIndex: index + 1 };
  }
  return undefined;
}

function liveSubscription(
  gateway: LiveGateway,
  headers: HeadersInit | undefined,
  topics: string[],
  options: { group: string; filter?: Filter },
): AsyncIterable<Message> {
  const startFrame = create(SubscribeFrameSchema, {
    frame: {
      case: "start",
      value: create(SubscribeStartSchema, {
        groupId: options.group,
        topics,
        autoCommit: true,
        filter: toProtoFilter(options.filter),
      }),
    },
  });
  return {
    [Symbol.asyncIterator]: () => liveSubscriptionIterator(gateway, headers, startFrame),
  };
}

function liveSubscriptionIterator(
  gateway: LiveGateway,
  headers: HeadersInit | undefined,
  startFrame: SubscribeFrame,
): AsyncIterator<Message> {
  const abortController = new AbortController();
  const inboundIterator = gateway
    .subscribe(openSubscribeFrames(startFrame), { headers, signal: abortController.signal })
    [Symbol.asyncIterator]();
  let isClosed = false;

  return {
    async next() {
      if (isClosed) {
        return closedMessageIteratorResult();
      }
      try {
        const inbound = await inboundIterator.next();
        if (isClosed || inbound.done) {
          return closedMessageIteratorResult();
        }
        return { done: false, value: fromProtoInbound(inbound.value) };
      } catch (error) {
        if (isClosed) {
          return closedMessageIteratorResult();
        }
        throw fromConnectError(error);
      }
    },
    return() {
      if (isClosed) {
        return Promise.resolve(closedMessageIteratorResult());
      }
      isClosed = true;
      abortController.abort();
      closeInboundIterator(inboundIterator);
      return Promise.resolve(closedMessageIteratorResult());
    },
  };
}

function closeInboundIterator(inboundIterator: AsyncIterator<Inbound>): void {
  if (!inboundIterator.return) {
    return;
  }
  void inboundIterator.return().catch(() => undefined);
}

function closedMessageIteratorResult(): IteratorResult<Message> {
  return { done: true, value: undefined };
}

function openSubscribeFrames(startFrame: SubscribeFrame): AsyncIterable<SubscribeFrame> {
  return {
    [Symbol.asyncIterator]: () => openSubscribeFrameIterator(startFrame),
  };
}

function openSubscribeFrameIterator(startFrame: SubscribeFrame): AsyncIterator<SubscribeFrame> {
  let startFrameWasSent = false;
  let closeControlStream: ((result: IteratorResult<SubscribeFrame>) => void) | undefined;

  return {
    next() {
      if (!startFrameWasSent) {
        startFrameWasSent = true;
        return Promise.resolve({ done: false, value: startFrame });
      }
      return new Promise<IteratorResult<SubscribeFrame>>((resolve) => {
        closeControlStream = resolve;
      });
    },
    return() {
      const result = { done: true, value: undefined } satisfies IteratorResult<SubscribeFrame>;
      closeControlStream?.(result);
      return Promise.resolve(result);
    },
    throw(error) {
      const result = { done: true, value: undefined } satisfies IteratorResult<SubscribeFrame>;
      closeControlStream?.(result);
      return Promise.reject(error);
    },
  };
}

function assertValidRecord(record: PublishRecord): void {
  if (record.topic === "") {
    throw new InvalidArgumentError("topic is required");
  }
  if (record.topic === "__missing_topic") {
    throw new NotFoundError("topic not found");
  }
}

function toProtoRecord(record: PublishRecord) {
  const headers = record.headers ?? [];
  return create(RecordSchema, {
    topic: record.topic,
    body: { case: "raw", value: record.value },
    headers: headers.map((header) => create(HeaderSchema, { key: header.name, value: header.value })),
  });
}

function toProtoFilter(filter: Filter | undefined): string {
  if (!filter) {
    return "";
  }
  const field = filter.path.startsWith("$.") ? filter.path.slice(2) : filter.path;
  if (typeof filter.value === "string") {
    return `${field} = '${filter.value.replaceAll("\\", "\\\\").replaceAll("'", "\\'")}'`;
  }
  return `${field} = ${filter.value}`;
}

function fromProtoInbound(inbound: Inbound): Message {
  return {
    topic: inbound.topic,
    partition: inbound.partition,
    offset: Number(inbound.offset),
    value: inbound.value,
    headers: inbound.headers.map((header) => ({ name: header.key, value: header.value })),
  };
}

function toMessage(stored: StoredMessage): Message {
  return {
    topic: stored.record.topic,
    partition: stored.partition,
    offset: stored.offset,
    value: new Uint8Array(stored.record.value),
    headers: cloneHeaders(stored.record.headers),
  };
}

function mockFilterMatches(filter: Filter | undefined, value: Uint8Array): boolean {
  if (!filter) {
    return true;
  }
  const field = filter.path.startsWith("$.") ? filter.path.slice(2) : undefined;
  if (!field) {
    return false;
  }
  try {
    const decoded = JSON.parse(new TextDecoder().decode(value)) as Record<string, unknown>;
    return decoded[field] === filter.value;
  } catch {
    return false;
  }
}

function cloneRecord(record: PublishRecord): Required<PublishRecord> {
  return { topic: record.topic, value: new Uint8Array(record.value), headers: cloneHeaders(record.headers ?? []) };
}

function cloneHeaders(headers: Header[]): Header[] {
  return headers.map((header) => ({ name: header.name, value: header.value ? new Uint8Array(header.value) : undefined }));
}
